### PR TITLE
A Row implementation that is based on ArrayObject which is faster

### PR DESCRIPTION
**Pull request by tsteur**
Currently, a `Row`'s columns are accessed like this: `$this->c[self::COLUMNS][$columnName]`. When archiving a big year instance we spend about 900 seconds (33%) just in `getColumn()`. By changing this to `$this->columns[$columnName]` the time dropped to 450 seconds (17%). By using `ArrayObject` the time drops by another 250 seconds. A difference in speed is not really noticeable in normal reports but when archiving aggregated reports (week, month, year, ranges).

Unfortunately, there is a huge disadvantage when using `ArrayObject`. It is not debuggable in PHPStorm and other IDE's as it is not supported by Xdebug see http://bugs.xdebug.org/view.php?id=996 and also kinda related http://bugs.xdebug.org/view.php?id=686 . Those issues are open since 2013 and 2011 so we can't expect a fix for this soon.

I'm issuing this pull request so we still have the code in case those bugs are resolved.

Please note in this PR there are also some changes to subtableId etc. which will be probably in master soon anyway. I didn't make the effort to split those changes as we only wanted to still have the code in general in case those bugs will be fixed in Xdebug. It wasn't easy to implement it based on `ArrayObject` since it implements the `Serializable` interface and we need to make sure to stay BC with already serialized Row instances in a different format see code.

### CODE REVIEW

File: core/DataTable.php [hunk 1]
```diff
@@ -1233,6 +1236,30 @@ public function getSerialized($maximumRowsInDataTable = null,
...
+    /**
+     * We cannot unserialize directly as it won't work to unserialize eg "Piwik\DataTable\Row" directly on
+     * PHP 5.6, 5.4.29 and 5.5.13. The Row class implements the "Serializable" interface and can therefore not be
+     * unserialized directly. This would trigger a notice and fail. Instead we use a class "Piwik_DataTable_SerializedRow"
+     * to unserialize which does not implement anything just to be able to access its "$row->c" property later.
+     * @param string $serialized
+     * @return array
+     * @throws Exception In case the unserialize fails
+     */
+    private function unserializeRows($serialized)
```
Comment (tsteur) [inline core/DataTable.php]: `ArrayObject` implements the `Serializable` interface therefore we cannot simple unserialize the serialized `Row` instances see comment. This was the tricky part in making it work with `ArrayObject`

File: core/DataTable.php [hunk 2]
```diff
@@ -1240,19 +1267,25 @@ public function getSerialized($maximumRowsInDataTable = null,
...
-        require_once PIWIK_INCLUDE_PATH . "/core/DataTable/Bridges.php";
+        $rows = $this->unserializeRows($serialized);
 
-        $serialized = unserialize($stringSerialized);
-        if ($serialized === false) {
-            throw new Exception("The unserialization has failed!");
+        foreach ($rows as $id => $row) {
+            if (isset($row->c)) {
+                $row = $row->c; // before Piwik 2.13
+            }
```
Comment (tsteur) [inline core/DataTable.php]: Pre Piwik 2.13 we have `Row::$c`, afterwards we just have an array of rows

File: core/DataTable/Row.php [hunk 3]
```diff
@@ -135,9 +92,10 @@ public function switchFlagSubtableIsLoaded()
      */
     public function __destruct()
     {
-        if ($this->isSubtableLoaded()) {
-            Manager::getInstance()->deleteTable($this->getIdSubDataTable());
-            $this->c[self::DATATABLE_ASSOCIATED] = null;
+        if ($this->isSubtableLoaded) {
```
Comment (tsteur) [inline core/DataTable/Row.php]: As mentioned in the PR the subtable changes are not really relevant to this PR

**Comment by mattab** [comment]
> It is not debuggable in PHPStorm and other IDE's as it is not supported by Xdebug see http://bugs.xdebug.org/view.php?id=996 

I've commented on this issue and also emailed Derick privately to offer to sponsor the work on this xdebug feature. I don't think there are other next steps for this task for now. 

this Pull request may be opened for a long time (ie. until Xdebug implements the feature).

**Comment by mattab** [comment]
Good news: following our request and sponsor offer, Derick from Xdebug has implemented the ArrayObject inspection feature and [the ticket](http://bugs.xdebug.org/view.php?id=996) was closed. Xdebug 2.3.3 is not yet released but should be in next weeks, then we can merge this PR. 

Moving to 2.14.0

**Comment by tsteur** [comment]
Actually, I will move this to Piwik 3.0 which we will work on "soon" anyway. Xdebug 2.3.3 is not released yet anyway and I also remember one issue with PHP 5.3 where the following is happening:

```
$row->setColumn('test', null);
$column = $row->getColumn('test');
// got: $column === null expected: $column === false
```

We could workaround this by using `$this->offsetExists($columnName)` instead of `isset($this[$columnName])` but it would make it a bit slower again...

File: core/DataTable/Filter/ColumnDelete.php [hunk 4]
```diff
@@ -134,9 +140,14 @@ public function filter($table)
                         && $name != 'label' // label cannot be removed via whitelisting
                         && !isset($this->columnsToKeep[$name])
                     ) {
-                        unset($table[$index][$name]);
+                        // we cannot remove row directly to prevent notice "ArrayIterator::next(): Array was modified
+                        // outside object and internal position is no longer valid in /var/www..."
+                        $columnsToDelete[] = $name;
                     }
                 }
```
Comment (fabiocarneiro) [inline core/DataTable/Filter/ColumnDelete.php]: add a empty line below this one

File: core/DataTable/Filter/ColumnDelete.php [hunk 5]
```diff
@@ -134,9 +140,14 @@ public function filter($table)
...
+                foreach ($columnsToDelete as $columnToDelete) {
```
Comment (fabiocarneiro) [inline core/DataTable/Filter/ColumnDelete.php]: add a empty line above this one

File: core/DataTable/Row.php [hunk 6]
```diff
@@ -73,60 +62,28 @@ class Row implements \ArrayAccess, \IteratorAggregate
...
-    /**
-     * Must be called after the row was serialized and __sleep was called.
-     * @ignore
-     */
-    public function cleanPostSerialize()
-    {
-        if ($this->subtableIdWasNegativeBeforeSerialize) {
-            $this->switchFlagSubtableIsLoaded();
-            $this->subtableIdWasNegativeBeforeSerialize = false;
+        if (isset($row[self::DATATABLE_ASSOCIATED])) {
```
Comment (fabiocarneiro) [inline core/DataTable/Row.php]: Use early return instead.

```php
if (isset($row[self::DATATABLE_ASSOCIATED])) {
    return;
}

if ($row[self::DATATABLE_ASSOCIATED] instanceof DataTable) {
    $this->setSubtable($row[self::DATATABLE_ASSOCIATED]);
    return;
}

$this->subtableId = $row[self::DATATABLE_ASSOCIATED];

**Comment by tsteur** [comment]
A new xdebug feature was released and we have a 3.0 branch. I will issue a new PR for this and we can merge it soon

**Pull request closed without merging**